### PR TITLE
fix(core): do not error when `ngDevMode` is undeclared

### LINKED ISSUE #31595: ReferenceError: ngDevMode is not defined
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core

### Is this a regression?

 Not sure

### Description

 A clear and concise description of the problem...
If there are compilation errors in a template (while using "ng serve"), I am seeing the error "ReferenceError: ngDevMode is not defined" when running in Firefox (60.6.1esr) or Edge. Chrome shows the actual compilation errors.

## 🔬 Minimal Reproduction

 https://stackblitz.com/...
Create simple app with a component. Introduce compilation errors. Try to run app in browser.

The error (with Angular 8.1.1 is in core.js:13393

## 🔥 Exception or Error
<pre><code>

ReferenceError: ngDevMode is not defined
</code></pre>

## 🌍  Your Environment

Angular Version: 8.1.1 (also 8.1.0)
<pre><code>

</code></pre>

**Anything else relevant?**

### PR DESCRIPTION
**Pull request by JoostK**
In production mode, the `ngDevMode` global may not have been declared.
This is typically not a problem, as optimizers should have removed all
usages of the `ngDevMode` variables. This does however require the
bundler/optimizer to have been configured in a certain way, as to allow
for `ngDevMode` guarded code to be removed.

As an example, Terser can be configured to remove the `ngDevMode`
guarded code using the following configuration:

```js
const terserOptions = {
  // ...
  compress: {
    // ...
    global_defs: require('@angular/compiler-cli').GLOBAL_DEFS_FOR_TERSER,
  }
}
```

(Taken from https://github.com/angular/angular/issues/31595#issuecomment-519129090)

If this is not done, however, the bundle should still work (albeit with
larger code size due to missed tree-shaking opportunities). This commit
adds a check for whether `ngDevMode` has been declared, as it is a
top-level statement that executes before `ngDevMode` has been initialized.

Fixes #31595

### CODE REVIEW

**Review by petebacondarwin — APPROVED** [review]

File: packages/core/src/core.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ export {Sanitizer} from './sanitization/sanitizer';
 export * from './codegen_private_exports';
 
 import {global} from './util/global';
-if (ngDevMode) {
+if (typeof ngDevMode !== 'undefined' && ngDevMode) {
```
Comment (crisbeto) [inline packages/core/src/core.ts:41]: Drive-by observation: the `typeof` check is something that can easily be missed. Maybe we can add a custom lint rule to check for it?

Comment (JoostK) [reply packages/core/src/core.ts:41]: I agree this is indeed error-prone and easy to regress in some way or another. There are however plenty of bare reads of `ngDevMode` in places that can depend on `initNgDevMode` to have been executed. A potential lint rule would therefore likely have to only consider top-level statements or something like that, which would also make it not entirely fool proof.

**Review by josephperrott — APPROVED** [review]
LGTM

**Review by atscott — APPROVED** [review]
